File: app/src/main/java/com/kooo/evcam/v2/plugin/V2StatusBarStateStore.java
package com.kooo.evcam.v2.plugin;

import android.content.Context;
import android.content.SharedPreferences;

public final class V2StatusBarStateStore {
    private static final String PREFS = "v2_status_bar_plugin";
    private static final String KEY_SERVICE_READY = "service_ready";
    private static final String KEY_RECORDING = "recording";
    private static final String KEY_EMERGENCY = "emergency";
    private static final String KEY_STATUS = "status";
    private static final String KEY_UPDATED_AT = "updated_at";

    private V2StatusBarStateStore() {
    }

    public static void update(Context context, boolean serviceReady, boolean recording, boolean emergency, String status) {
        prefs(context).edit()
                .putBoolean(KEY_SERVICE_READY, serviceReady)
                .putBoolean(KEY_RECORDING, recording)
                .putBoolean(KEY_EMERGENCY, emergency)
                .putString(KEY_STATUS, status == null ? "" : status)
                .putLong(KEY_UPDATED_AT, System.currentTimeMillis())
                .apply();
    }

    public static Snapshot read(Context context) {
        if (context == null) {
            return Snapshot.empty();
        }
        SharedPreferences prefs = prefs(context);
        return new Snapshot(
                prefs.getBoolean(KEY_SERVICE_READY, false),
                prefs.getBoolean(KEY_RECORDING, false),
                prefs.getBoolean(KEY_EMERGENCY, false),
                prefs.getString(KEY_STATUS, "")
        );
    }

    private static SharedPreferences prefs(Context context) {
        return context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public static final class Snapshot {
        public final boolean serviceReady;
        public final boolean recording;
        public final boolean emergency;
        public final String status;

        private Snapshot(boolean serviceReady, boolean recording, boolean emergency, String status) {
            this.serviceReady = serviceReady;
            this.recording = recording;
            this.emergency = emergency;
            this.status = status == null ? "" : status;
        }

        private static Snapshot empty() {
            return new Snapshot(false, false, false, "");
        }
    }
}

File: app/src/main/kotlin/com/kooo/evcam/v2/plugin/V2StatusBarStateStore.kt
package com.kooo.evcam.v2.plugin

import android.content.Context

object V2StatusBarStateStore {
    private const val PREFS = "v2_status_bar_plugin"
    private const val KEY_SERVICE_READY = "service_ready"
    private const val KEY_RECORDING = "recording"
    private const val KEY_STATUS = "status"
    private const val KEY_UPDATED_AT = "updated_at"

    @JvmStatic
    fun update(context: Context, serviceReady: Boolean, recording: Boolean, status: String?) {
        prefs(context).edit()
            .putBoolean(KEY_SERVICE_READY, serviceReady)
            .putBoolean(KEY_RECORDING, recording)
            .putString(KEY_STATUS, status.orEmpty())
            .putLong(KEY_UPDATED_AT, System.currentTimeMillis())
            .commit()
    }

    @JvmStatic
    fun read(context: Context?): Snapshot {
        if (context == null) return Snapshot.empty()
        val prefs = prefs(context)
        return Snapshot(
            serviceReady = prefs.getBoolean(KEY_SERVICE_READY, false),
            recording = prefs.getBoolean(KEY_RECORDING, false),
            status = prefs.getString(KEY_STATUS, "").orEmpty(),
        )
    }

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    data class Snapshot(
        @JvmField val serviceReady: Boolean,
        @JvmField val recording: Boolean,
        @JvmField val status: String,
    ) {
        companion object {
            fun empty() = Snapshot(false, false, "")
        }
    }
}
